Adding Diff Trimmed Lines

JamesGould123 opened the pull request:
I extended the LineDiff.tokenize function to add the ability to ignore leading and trailing spaces when using TrimmedLineDiff, and added diffTrimmedLines to the public-facing functions. My changes should not affect any previous functionality.

JamesGould123 commented:
I've realized that I did not add documentation in the README, I can do this as well once this pull request is merged.

kpdecker commented:
@JamesGould123 you can push additional commits to the branch with any changes that you'd like to make. Docs and tests are something I'd like to see before merging.

kpdecker commented on diff.js:
```diff
@@ -114,8 +115,8 @@
-          newString = this.tokenize(newString);
-          oldString = this.tokenize(oldString);
+          newString = this.tokenize(newString, !!self.ignoreTrim);
+          oldString = this.tokenize(oldString, !!self.ignoreTrim);
```
Rather than changing the tokenize API, can't we just read this flag below?

JamesGould123 replied:
Oh, certainly, thanks for the line note. I worked myself into a logical corner, and couldn't think of any other way to do that when I was coding, so I did this hesitantly. Looking at it again, the correct solution is quite clear.

JamesGould123 commented:
@kpdecker Sorry, I'm new to using Github, didn't know I could edit a pull request I've already sent, I'll add documentation and look into adding tests tonight.

kpdecker commented:
@JamesGould123 no worries. Welcome to github! :)

Just so you are aware, github does not send updates when the PR changes, so you should make a comment here saying that you've made the changes so I'll get a notification.

JamesGould123 commented:
@kpdecker Thanks!

Okay, I've overcome my fear of npm, added tests, moved the edits I made to the tokenize API into the LineDiff.tokenize function, and added documentation.

kpdecker commented on README.md:
```diff
@@ -33,6 +33,10 @@ or
 
     Returns a list of change objects (See below).
 
+* `JsDiff.TrimmedLineDiff(oldStr, newStr[, callback])` - diffs two blocks of text, comparing line by line, ignoring leading and trailing whitespace.
```
This should be `JsDiff.diffTrimmedLines`

JamesGould123 replied:
Sorry, my bad.

kpdecker commented on test/diffTest.js:
```diff
@@ -200,6 +200,36 @@ describe('#diffLines', function() {
+    diff.convertChangesToXML(diffResult).should.equal('line\nvalue\nline');
+  });
+
+  it('should ignore shorespace', function() {
```
Could you say something other than shorespace? This isn't really clear what the test is intending.

JamesGould123 replied:
On line 188, in the tests for #diffLines:

<code>it('should handle shorespace', function() {</code>

I was reversing that test. Shorespace refers to the leading and trailing whitespace. I think both tests should be titled similarly, so they can be easily juxtaposed. Should I switch both to say "leading and trailing whitespace" instead of "shorespace"?

kpdecker replied:
I think leading and trailing whitespace is a bit clearer. I didn't know what shorespace meant when I read it.

JamesGould123 replied:
Agreed, I'll change that.

kpdecker commented on diff.js:
```diff
@@ -256,18 +257,34 @@
         // Merge lines that may contain windows new lines
-        if (line === '\n' && lastLine && lastLine[lastLine.length - 1] === '\r') {
-          retLines[retLines.length - 1] += '\n';
+        if (line === '\n' && lastLine && lastLine[lastLine.length - 1]) {
```
The removal of the `'\r'` check seems like it will break something. Why was this removed?

JamesGould123 replied:
Huh, I'm surprised that didn't fail any of the tests. It passes all of the tests either way, but I'll re-add the "=== '\r'" because it should definitely be there. Maybe I should add a more verbose test for how it handles newlines.

kpdecker commented on diff.js:
```diff
@@ -256,18 +257,34 @@
-          retLines[retLines.length - 1] += '\n';
+        if (line === '\n' && lastLine && lastLine[lastLine.length - 1]) {
+          if(this.ignoreTrim){
+            retLines[retLines.length - 1] = retLines[retLines.length - 1].slice(0,-1) + '\r\n';
```
Why is this `\r\n` here but `\n` below?

JamesGould123 replied:
string.trim() in JS removes the newline characters, which puts the entire diff on one line, so I added the '\n' character after trimming every line but the last line. However, then when it tried to merge empty lines using:

<code>retLines[retLines.length - 1] += '\n';</code>

It would end up with '\n\n' on the end of the line, instead of '\r\n' I fixed this by slicing and re-adding the '\r\n'.

JamesGould123 commented:
Okay, I just pushed a commit fixing the documentation issue and extending the merge lines functionality and adding a comment explaining it.

JamesGould123 commented:
@kpdecker Have you looked at the most recent changes? I've done everything you requested other than changing the test, which I explained and asked your opinion on above.

kpdecker commented on diff.js:
```diff
@@ -256,18 +257,36 @@
-          retLines[retLines.length - 1] += '\n';
+        if (line === '\n' && lastLine &&
+          (lastLineLastChar === '\r' || lastLineLastChar === '\n')) {
+          if(this.ignoreTrim || lastLineLastChar === '\n'){
```
Please match whitespace style. (Space after `if` and before `{`, no `else` on the same line)

JamesGould123 replied:
I think you mean else should be on on the same line, it appears to be same line throughout the rest of the file, like this:
<pre>
if () {

} else {

}</pre>

JamesGould123 replied:
There are several places in applyPatch where these rules are not followed, I'll fix that as well.

kpdecker commented on diff.js:
```diff
@@ -256,18 +257,36 @@
         // Merge lines that may contain windows new lines
-        if (line === '\n' && lastLine && lastLine[lastLine.length - 1] === '\r') {
-          retLines[retLines.length - 1] += '\n';
+        if (line === '\n' && lastLine &&
```
This `lastLine` check isn't necessary since it was there to prevent a `undefined is not an object` error when reading from the array and that is now done when initializing the `lastLineLastChar` variable..

JamesGould123 replied:
Nice catch.

kpdecker commented on diff.js:
```diff
@@ -256,18 +257,36 @@
+          if(this.ignoreTrim){
+            line = line.trim();
+            //add a newline unless this is the last line.
+            if(!(i + 1 === lines.length)){
```
nit pick: `if (i < lines.length - 1) {` save a few characters when minimized (or `i + 1`, the key thing is using the `<` operator rather than `!` and `===`)

kpdecker commented on diff.js:
```diff
@@ -256,18 +257,36 @@
+          (lastLineLastChar === '\r' || lastLineLastChar === '\n')) {
+          if(this.ignoreTrim || lastLineLastChar === '\n'){
+            //to avoid merging to \n\n, remove \n and add \r\n.
+            retLines[retLines.length - 1] = retLines[retLines.length - 1].slice(0,-1) + '\r\n';
```
I don't think we should use `\r\n` here. We are normalizing it to `\n` for all other lines so it's surprising to have it here and nowhere else. Instead, I would do nothing. I.e.

```
if (line === '\n' && lastLine && lastLine[lastLine.length - 1] === '\r') {
  if (!this.ignoreTrim) {
    retLines[retLines.length - 1] += '\n';
  }
}
```

We can do this because we know that the line will always have a `\n` at the end due to the `+=` operation below.

JamesGould123 replied:
This causes the tests I wrote to fail, which isn't necessarily a problem, but I followed the test set by the Line Diff tests, so I don't think it is exhibiting different behavior that that.

The final Line Diff test is:

<pre>
it('should handle windows line endings', function() {
    var diffResult = diff.diffLines(
      'line\r\nold value \r\nline',
      'line\r\nnew value\r\nline');
    diff.convertChangesToXML(diffResult).should.equal('line\r\n<ins>new value\r\n</ins><del>old value \r\n</del>line');
  });
</pre>

The final Trimmed Line Diff test is:

<pre>
  it('should handle windows line endings', function() {
    var diffResult = diff.diffTrimmedLines(
      'line\r\nold value \r\nline',
      'line\r\nnew value\r\nline');
    diff.convertChangesToXML(diffResult).should.equal('line\r\n<ins>new value\r\n</ins><del>old value\r\n</del>line');
  });
</pre>

Changing to your suggested code causes the second test to fail with the following value (no \r):

<code>'line\n<ins>new value\n</ins><del>old value\n</del
>line'</code>

This is the only thing I'm not sure about, everything else should be updated in my most recent commit.

kpdecker commented:
A side note, I'm leaving for vacation shortly so my responses might take a bit longer, but I do want to get this merged!

JamesGould123 commented:
@kpdecker, okay I've fixed everything except that last issue. Have a good time on your vacation!

kpdecker commented:
Released in 1.3.0

JamesGould123 commented:
Thanks :smile:
